dev-tools/xtask/src/download.rs

```diff
@@ -0,0 +1,935 @@
+        unpack_tarball(&self.log, &tarball_path, &download_dir).await?;
+
+        let _ = tokio::fs::remove_dir_all(&destination_dir).await;
+        tokio::fs::create_dir_all(&destination_dir).await?;
+        let destination_root = destination_dir.join("root");
+        tokio::fs::create_dir_all(&destination_root).await?;
+        copy_dir_all(&download_dir.join("root"), &destination_root)?;
+
+        let bin_dir = destination_dir.join("root/opt/oxide/dendrite/bin");
+
+        // Symbolic links for backwards compatibility with existing setups
+        std::os::unix::fs::symlink(
+            bin_dir.canonicalize()?,
+            destination_dir.canonicalize()?.join("bin"),
+        )
+        .context("Failed to create a symlink to dendrite's bin directory")?;
+
+        match os_name()? {
+            Os::Linux => {
+                let base_url =
+                    format!("{BUILDOMAT_URL}/{repo}/linux-bin/{commit}");
+                let filename = "dpd";
+                let path = download_dir.join(filename);
+                download_file_and_verify(
+                    &self.log,
+                    &path,
+                    &format!("{base_url}/{filename}"),
+                    ChecksumAlgorithm::Sha2,
+                    &dpd_sha2,
+                )
+                .await?;
+                set_permissions(&path, 0o755).await?;
+                tokio::fs::copy(path, bin_dir.join(filename)).await?;
+
+                let filename = "swadm";
+                let path = download_dir.join(filename);
+                download_file_and_verify(
+                    &self.log,
+                    &path,
+                    &format!("{base_url}/{filename}"),
+                    ChecksumAlgorithm::Sha2,
+                    &swadm_sha2,
+                )
+                .await?;
+                set_permissions(&path, 0o755).await?;
+                tokio::fs::copy(path, bin_dir.join(filename)).await?;
+            }
+            Os::Illumos => {}
+            Os::Mac => {
+                warn!(self.log, "WARNING: Dendrite not available for Mac");
+                warn!(self.log, "Network APIs will be unavailable");
+
+                let path = bin_dir.join("dpd");
+                tokio::fs::write(&path, "echo 'unsupported os' && exit 1")
+                    .await?;
+                set_permissions(&path, 0o755).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn download_maghemite_mgd(&self) -> Result<()> {
+        let download_dir = self.output_dir.join("downloads");
+        tokio::fs::create_dir_all(&download_dir).await?;
+
+        let checksums_path = self.versions_dir.join("maghemite_mgd_checksums");
+        let [mgd_sha2, mgd_linux_sha2] = get_values_from_file(
+            ["CIDL_SHA256", "MGD_LINUX_SHA256"],
+            &checksums_path,
+        )
+        .await?;
+        let commit_path =
+            self.versions_dir.join("maghemite_mg_openapi_version");
```

Comment:
Using the [rev value from Cargo.toml](https://github.com/oxidecomputer/omicron/blob/dbf849401c08c62bfcc64c28292f048469b10ef8/Cargo.toml#L273) is fine here, and we can probably get rid of the `maghemite_mg_openapi_version` file.

Reply:
I agree that we probably should do this, but it's slightly sticky because the `flake.nix` file also depends on having this as a rev.

I agree that we should minimize version dups - but if it's okay with you, I'm going to punt this to a different PR, since my goal is "parity with the bash scripts" before changing this download process **too** much.

Reply:
The Nix flake could be rewritten to also parse the rev from `Cargo.toml`, FWIW, but I would love to have advance warning so I can make sure to actually do that. :)

Reply:
SGTM